### DISCUSSION TITLE
Import mechanism does not reduce load on LDAP

### DISCUSSION THREAD

**Discussion by AlexanderSkrock** [keycloak/keycloak#23116] [Q&A]
Hi!

I recently set up a keycloak server with LDAP as user federation prodivder. We enabled the import mechanism. With this configuration I hoped that this will reduce the load on our LDAP by a lot. In theory it could be minimized to only verify passwords.

But, actually the opposit happens. When our application queries for users, the query is always executed against keycloak's local storage and all user federation provider. After query execution there further is some "import validation" per code, which against loads the data for each user in the query result from LDAP.
In the beginning there also have been additional ldap requests for groups and roles but we were able to eliminate those by toggling the import flag groups and roles.

Is there some configuration mistake I could have done? Also, what is the intended behaviour?

I appreciate any help or hints on this!

Kind regards
Alexander Skrock

Update:
The `LDAPStorageProvider` implements the `ImportedUserValidation` interface and thus, every user that is retrieved by querying is validated against this provider. This leads to one ldap request per user. E. g. if I query users by first name and have 100 results, the we have 100 additional request, one person, to retrieve its attributes and validate the imported attributes. This happens because the local users has a federation link.

**Comment by darius-m** [comment]
I think you at least need to disable the setting that requires some attributes to always be queried in the back-end in the federation provider's mappers. I think the setting may have disappeared from the current admin interface however, so you may need to update the database directly. Honestly not sure how to find the setting currently; if you export the realm (or more specifically the federation provider), you will see that there is a setting in the JSON file for each mapper - theoretically you could also export the provider, alter the attributes and then import it back. Not entirely sure if your issue is caused by this.

**Comment by AlexanderSkrock** [comment]
Thank you @darius-m for having a look into my problem. I also stumbled across this flag and double-checked it is disabled for any of the defined mappers. Sadly this is the case, so it cannot cause my issue.

Kind regards
Alexander Skrock
